ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HivePointLookupOptimizerRule.java

```diff
@@ -727,44 +708,58 @@ private RexNode useStructIfNeeded(List<? extends RexNode> columns) {
                 }
               } else {
                 for (int j = 1; j < inCall.getOperands().size(); j++) {
-                  String expr = inCall.getOperands().get(j).toString();
-                  inLHSExprToRHSExprs.put(ref, expr);
-                  stringToExpr.put(expr, inCall.getOperands().get(j));
+                  inLHSExprToRHSExprs.put(ref, inCall.getOperands().get(j));
+                }
+              }
+              operands.remove(i);
+              --i;
+            } else if (operand.getKind() == SqlKind.EQUALS) {
+              Constraint c = Constraint.of(operand);
+              if (c == null || !HiveCalciteUtil.isDeterministic(c.exprNode)) {
+                continue;
+              }
+              RexNode ref = c.exprNode;
+              if (inLHSExprToRHSExprs.containsKey(ref)) {
+                inLHSExprToRHSExprs.get(ref).retainAll(Collections.singleton(c.constNode));
+                if (!inLHSExprToRHSExprs.containsKey(ref)) {
+                  // Note that Multimap does not keep a key if all its values are removed.
+                  // Hence, since there are no common expressions and it is within an AND,
+                  // we should return false
+                  return rexBuilder.makeLiteral(false);
```

Comment:
I think this may not neccessarily false; consider:
```
x=1 and x=2
```
which is `not true`; but in case `x is null` - the result of the `AND` will be `null`

instead of false this could be only rewritten to `x is null and null`; but even in that case there could be other stuff which might affect null/false;
so I think we may not do a return here.

Reply:
Good catch. It seems the problem was already there though. I am revisiting this logic.